### PR TITLE
Add third-party skeletal features

### PR DESCRIPTION
**Pull request by karthikprabhuvinod**
This PR starts the third-party repo-add, delete, list feature
for swupd. This PR also add support for configure flag
to build swupd optionally, off by default.

**Repo Add:**
 sudo swupd 3rd-party add <repo_name> \<url\>
**Repo Remove:**
 sudo swupd 3rd-party remove <repo_name>
**Repos List:**
 sudo swupd 3rd-party list

Signed-off-by: Karthik Prabhu Vinod <karthik.prabhu.vinod@intel.com>

### CODE REVIEW

**Comment by karthikprabhuvinod** [comment]
@otaviobp : I have a working third_party_repo PR...Tests are still pending and PR needs a bit of finishing. At this moment, could you  take a look from an early feedback & feature standpoint ?

File: Makefile.am [hunk 1]
```diff
@@ -32,6 +32,11 @@ bin_PROGRAMS = swupd
 AM_CPPFLAGS = $(AM_CFLAGS) $(libarchive_CFLAGS)
 
 swupd_SOURCES = \
+	src/3rd_party_add.c \
+	src/3rd_party.c \
+	src/3rd_party_helpers.c \
+        src/3rd_party_list.c \
```
Comment (otaviobp) [inline Makefile.am]: mix tab/spaces?

File: configure.ac [hunk 2]
```diff
@@ -72,6 +72,25 @@ AS_IF(
   [SIGVERIFICATION="no"]
 )
 
+AC_ARG_ENABLE(
+  [thirdparty],
```
Comment (otaviobp) [inline configure.ac]: I prefer third-party, but I'm ok with all together too.

File: configure.ac [hunk 3]
```diff
@@ -72,6 +72,25 @@ AS_IF(
...
+)
+THIRDPARTY="no"
+AS_IF(
+  [test "x$enable_thirdparty" = "xyes"],
+  [AC_DEFINE(THIRDPARTY, 1, [Enable third party])
+     THIRDPARTY="yes"],
+  [THIRDPARTY="no"]
+)
+
+AH_TEMPLATE([THIRDPARTY_REPO_PREFIX], [prefix path for third party repos])
```
Comment (otaviobp) [inline configure.ac]: ++ for having a setting for REPO_PREFIX

File: src/3rd_party.c [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+#define CHOOSE_REMOVE 9003
+static bool cmdline_option_add;
+static bool cmdline_option_list;
+static bool cmdline_option_remove;
+
+#ifdef THIRDPARTY
+
+void get_repo_path(char **repo_path, char *repo_name)
+{
+	string_or_die(repo_path, "%s%s", THIRDPARTY_REPO_PREFIX, repo_name);
```
Comment (otaviobp) [inline src/3rd_party.c]: Use sys_path_join here to make sure we are going to have a '/' separator

File: src/3rd_party.c [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+static void print_help(void)
+{
+	print("Usage:\n");
+	print("   swupd 3rd-party [OPTION...]\n\n");
+
+	global_print_help();
+
+	print("Options:\n");
+	print("   --add <repo_name>		Add third party repo\n");
```
Comment (otaviobp) [inline src/3rd_party.c]: I don't think it's a good idea to use flags for add, list remove. If you use a flag we need to protect against multiple flags used together and this will become hard after adding bundle-add and other commands

File: src/3rd_party.c [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+	progress_finish_steps(ret);
+	free_string(&repo_path);
+	return ret;
+}
+
+#else
+
+enum swupd_code third_party_main(int UNUSED_PARAM argc, char UNUSED_PARAM **argv)
+{
+	info("This feature is disabled at compile time\n");
```
Comment (otaviobp) [inline src/3rd_party.c]: It's better to avoid a print here so for users this feature won't exist before it's released officially.

File: src/3rd_party.h [hunk 7]
```diff
@@ -0,0 +1,8 @@
+#ifndef __THIRD_PARTY_H
+#define __THIRD_PARTY_H
+
```
Comment (otaviobp) [inline src/3rd_party.h]: I would add the `#ifdef THIRDPARTY` here and never call those functions if it's false.

Comment (otaviobp) [reply src/3rd_party.h]: And follow the standard in other header files to create this one. The define name is different and there's `extern "C" `missing

File: swupd.bash [hunk 8]
```diff
@@ -39,7 +39,7 @@ _swupd()
 	    ("check-update")
 		opts="$global "
 		break;;
-		("update")
+	    ("update")
```
Comment (otaviobp) [inline swupd.bash]: tab/spaces?

File: src/3rd_party.c [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+	int ind = global_parse_options(argc, argv, &opts);
+	int opt_mask = 0;
+
+	if (ind < 0) {
+		return false;
+	}
+
+	if (!cmdline_option_repo) {
+		if (cmdline_option_add || cmdline_option_remove) {
+			error("The --repo option is a required argument\n\n");
```
Comment (otaviobp) [inline src/3rd_party.c]: same for --repo. I prefer:

`swupd 3rd-party add repo_name url`

File: src/3rd_party_add.c [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+		error("failed to create %s\n", repo_path);
+		return -1;
+	}
+
+	return SWUPD_OK;
+}
+
+int third_party_add(char *repo_path)
+{
+	if (repo_exists(repo_path)) {
```
Comment (otaviobp) [inline src/3rd_party_add.c]: We need to have a file in /var/lib/swupd to track the 3rd-party repositories and the corresponding url, so checking for the dir is not enough. And if the dir is completly removed we should be able to restore that.

File: src/3rd_party.c (hunk 5 above)
Comment (karthikprabhuvinod) [reply src/3rd_party.c]: > I don't think it's a good idea to use flags for add, list remove. If you use a flag we need to protect against multiple flags used together and this will become hard after adding bundle-add and other commands

I was thinking about that....I need to have add, list, remove as sub_commands like add in git remote add

File: src/3rd_party.c (hunk 6 above)
Comment (karthikprabhuvinod) [reply src/3rd_party.c]: Makes sense will change in next push

File: Makefile.am (hunk 1 above)
Comment (karthikprabhuvinod) [reply Makefile.am]: Makes sense will change in next push

File: src/3rd_party.h (hunk 7 above)
Comment (karthikprabhuvinod) [reply src/3rd_party.h]: > I would add the `#ifdef THIRDPARTY` here and never call those functions if it's false.

I think we are using the same pattern using compile time flag for signatures...like in c src/signature.c

Comment (karthikprabhuvinod) [reply src/3rd_party.h]: > And follow the standard in other header files to create this one. The define name is different and there's `extern "C" `missing

Makes sense will change in next push

Comment (otaviobp) [reply src/3rd_party.h]: I mixed 2 comments in same line and I realized now that it's confusing.
 - The #ifdef style is good the way it is, like the signature as you said. But If you just ignore all file it's going to be easier to code and I'm ok with that. But if you want to leave like it is, it's fine too.
 - The style I was talking about is the header file. But I realized that there's not a very well established style. each file does it in a different way. So the only thing really missing is the extern.

File: src/main.c [hunk 11]
```diff
@@ -69,6 +69,7 @@ static struct subcmd commands[] = {
 	{ "clean", "Clean cached files", clean_main },
 	{ "hashdump", "Dump the HMAC hash of a file", hashdump_main },
 	{ "verify", "NOTE: this command has been superseded, please use \"swupd diagnose\" instead", diagnose_main },
+	{ "3rd-party", "Indicate Swupd to use user's third party repo", third_party_main },
```
Comment (otaviobp) [inline src/main.c]: And use the ifdef here too. I forgot to mention that in my first review

File: src/3rd_party.c (hunk 9 above)
Comment (karthikprabhuvinod) [reply src/3rd_party.c]: Actually I was thinking this should be an option for sure. Usually command are actions.

It works perfectly in our case as then we can make it required for remove, add and not have it for list.

Let me know what you feel

File: src/3rd_party.h (hunk 7 above)
Comment (karthikprabhuvinod) [reply src/3rd_party.h]: For point1 I will try to do what you said. I agree with you
For point2, will make sure headers have extern c

File: src/main.c (hunk 11 above)
Comment (karthikprabhuvinod) [reply src/main.c]: Ahh yes.. will do

File: src/3rd_party.c (hunk 4 above)
Comment (karthikprabhuvinod) [reply src/3rd_party.c]: Makes sense will change in next push

File: src/3rd_party.c (hunk 5 above)
Comment (karthikprabhuvinod) [reply src/3rd_party.c]: Makes sense will change in next push

File: src/3rd_party.h (hunk 7 above)
Comment (karthikprabhuvinod) [reply src/3rd_party.h]: Makes sense will change in next push

File: swupd.bash (hunk 8 above)
Comment (karthikprabhuvinod) [reply swupd.bash]: Makes sense will change in next push

File: src/main.c (hunk 11 above)
Comment (karthikprabhuvinod) [reply src/main.c]: Makes sense will change in next push

**Comment by karthikprabhuvinod** [comment]
@otaviobp 
I have addressed all your comments with remarks like "Makes sense will change in next push" except two i guess.

With this PR, I made sure the skeleton exists, The add, delete list command lines parsers are created. But they don't do much. 

Please could give it a second quick look? Anything you want to be removed for now?

File: src/3rd_party.h [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+#include "swupd.h"
+
+/**
+ * @brief Creates a repo folder under THIRDPARTY_REPO_PREFIX 
+ * if it does not exist
+ *
+ * @return 0 on success, SWUPD_THIRD_PARTY_REPO_EXISTS if it already exists
+ * and negative numbers on errors
+ */
+enum swupd_code third_party_add(int argc, char **argv);
```
Comment (otaviobp) [inline src/3rd_party.h]: Rename all functions that are a main to add a _main suffix. This one should be `third_party_add_main`

File: src/3rd_party.c [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+};
+
+static struct subcmd commands[] = {
+	{ "add", "Add third party repo", third_party_add },
+	{ "remove", "Remove third party repo", third_party_remove },
+	{ "list", "List third party repo", third_party_list },
+	{ 0 }
+};
+
+static int subcmd_index(char *arg)
```
Comment (otaviobp) [inline src/3rd_party.c]: Instead of copying, export this function in swupd_internal.h and use it in 3rd_party

Comment (karthikprabhuvinod) [reply src/3rd_party.c]: @otaviobp This command is slightly diffferent than the one we used in main.c since we dont have to check if it is a verify command etc. Should I still use that same function? Also I will need to pass commands args to it since commands are different from main.c, add.c remove.c

File: Makefile.am [hunk 14]
```diff
@@ -32,6 +32,10 @@ bin_PROGRAMS = swupd
 AM_CPPFLAGS = $(AM_CFLAGS) $(libarchive_CFLAGS)
 
 swupd_SOURCES = \
+	src/3rd_party_add.c \
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply Makefile.am]: Makes sense will change in next push

File: src/3rd_party.h [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+/**
+  * @file
+  * @brief Functions to call for third_party_main
+  */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "swupd.h"
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party.h]: Makes sense will change in next push

File: src/3rd_party.h [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * @brief Creates a repo folder under THIRDPARTY_REPO_PREFIX
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party.h]: Makes sense will change in next push

File: src/3rd_party_add.c [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+#include "lib/log.h"
+#include "swupd.h"
+#include "sys/stat.h"
+
+#ifdef THIRDPARTY
+
+static int create_repo_dir(char UNUSED_PARAM *repo_path)
+{
+	//repo_path doesn't exist
+	/*
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: Makes sense will change in next push

File: src/3rd_party_internal.h [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+	return SWUPD_OK;
+}
+
+static inline int setup_repo_tracking()
+{
+	char *tracking_dir;
+	char *tracking_file;
+	int ret;
+	tracking_dir = sys_path_join(globals.path_prefix, globals.state_dir);
+	tracking_file = sys_path_join(tracking_dir, "repo_tracker.ini");
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_internal.h]: Makes sense will change in next push

File: src/swupd_exit_codes.h [hunk 19]
```diff
@@ -47,6 +47,8 @@ enum swupd_code {
 	SWUPD_OUT_OF_MEMORY_ERROR,	     /* 34 swupd ran out of memory */
 	SWUPD_VERIFY_FAILED,		     /* 35 verify could not fix/replace/delete one or more files */
 	SWUPD_INVALID_BINARY,		     /* 36 binary to be executed is missing or invalid */
+	SWUPD_THIRD_PARTY_REPO_EXISTS,	     /* 37 third party repo already exists, error for adding */
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/swupd_exit_codes.h]: Makes sense will change in next push

File: src/3rd_party.h (hunk 12 above)
Comment (karthikprabhuvinod) [reply src/3rd_party.h]: Makes sense will change in next push

File: src/3rd_party_internal.h [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * @brief Check if a repo directory exists
+ *
+ * Check if the directory pointed to repo_exists exists
+ *
+ * @return true on success, false on failure
+ */
+static inline bool repo_exists(char *repo_path)
+{
+	struct stat sb;
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_internal.h]: Makes sense will change in next push

File: src/3rd_party_internal.h [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+
+	if ((sb.st_uid == 0) &&
+	    (S_ISDIR(sb.st_mode)) &&
+	    ((sb.st_mode & 0777) == 0700)) {
+		return true;
+	}
+	return false;
+}
+
+static int create_tracking_file(char *tracking_file)
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_internal.h]: @otaviobp : for me, a name like create_repo makes more sense with something like creating a repo directory. Here we are talking about creating a tracking file.

[the remark this answers is not in the document]
Comment (otaviobp) [reply src/3rd_party_internal.h]: So probably I didn't understand the usage of this function. What's a tracking file?

File: src/3rd_party.c (hunk 13 above)
Comment (otaviobp) [reply src/3rd_party.c]: take a look at  #1180
And you can get that as a parameters. You can rename that if you find a better name and add a comment explaining what the function is doing.

File: src/3rd_party_internal.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_internal.h]: Its the repo.ini file where write what repo it is and all the bundles under it

File: src/3rd_party.c (hunk 13 above)
Comment (karthikprabhuvinod) [reply src/3rd_party.c]: Okay Cool I will move to swupd_internal.h as a inline and include it in main.c and 3rd-party.c

Comment (karthikprabhuvinod) [reply src/3rd_party.c]: Makes sense will change in next push

File: src/3rd_party_internal.h (hunk 21 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_internal.h]: Makes sense will change in next push

File: src/swupd_internal.h [hunk 22]
```diff
@@ -17,5 +17,36 @@ extern enum swupd_code install_main(int argc, char **argv);
...
+	enum swupd_code (*mainfunc)(int, char **);
+};
+
+static inline int subcmd_index(char *arg, struct subcmd *commands)
+{
+	struct subcmd *entry = commands;
+	int i = 0;
+	size_t input_len;
+	size_t cmd_len;
+
```
Comment (castulo) [inline src/swupd_internal.h]: why did you remove this from the function?
```
if (strcmp(arg, "verify") == 0) {		
 	verify_set_command_verify(true); // set to true so we know the "verify" command was used		
 	arg = "diagnose";		
 }
```

File: swupd.zsh [hunk 23]
```diff
@@ -175,14 +175,41 @@ if [[ -n "$state" ]]; then
         "mirror:Configure mirror url for swupd content"
         "clean:Clean cached files"
         "hashdump:Dump the HMAC hash of a file"
+	"3rd-party: Indicate Swupd to use user's third party repo"
```
Comment (castulo) [inline swupd.zsh]: you seem to be mixing tabs and spaces in these files

File: src/swupd_internal.h (hunk 22 above)
Comment (otaviobp) [reply src/swupd_internal.h]: see #1180

Comment (castulo) [reply src/swupd_internal.h]: just did, that explains it, thanks

**Comment by otaviobp** [comment]
> This PR seems to be moving in the right direction but there are still many things that need to be addressed first. I added comments with all my observations in the PR. I also noted there are no tests whatsoever for this PR, it introduces lots of new functionality so I think having tests is crucial.

This PR isn't really adding a new functionality, so there's no way on creating tests to it. It's just adding the skeleton so we can move forward.

**Comment by otaviobp** [comment]
@karthikprabhuvinod Can you remove all functions from this PR that aren't related to the skeleton? And you can add later when you really need them.

File: src/3rd_party.c [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+#include "3rd_party.h"
+#include "config.h"
+#include "lib/log.h"
+#include "swupd.h"
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party.c]: Makes sense will remove it in next push

File: src/3rd_party_add.c [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+#include "3rd_party_internal.h"
+#include "config.h"
+#include "lib/log.h"
+#include "swupd.h"
+#include "sys/stat.h"
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: There are some others which are redundant as well. Makes sense will remove it in next push

File: src/3rd_party_add.c [hunk 26]
```diff
@@ -0,0 +1,119 @@
...
+		return false;
+	}
+
+	if (argc == 1) {
+		error("The positional args: repo-name and URL are missing\n\n");
+		return false;
+	} else if (argc == 2) {
+		error("The positional args: repo-URL is missing\n\n");
+		return false;
+	}
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: Good point.  but Actually am not expecting any more args as such for these commands.

File: src/3rd_party_add.c [hunk 27]
```diff
@@ -0,0 +1,119 @@
...
+	if (ret != SWUPD_OK) {
+		goto finish;
+	}
+
+	repo_path = get_repo_path(argv[1]);
+
+	ret = create_repo_dir(repo_path);
+
+	if (!ret) {
+		info("Third party repo successfully created\n");
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: @castulo Very sharp observation. Yes will change it.

File: src/3rd_party_add.c [hunk 28]
```diff
@@ -0,0 +1,119 @@
...
+		create_repo(argv[1]);
+		goto finish;
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: Makes sense will remove it in next push

File: src/3rd_party_add.c [hunk 29]
```diff
@@ -0,0 +1,119 @@
...
+	progress_finish_steps(ret);
+	free_string(&repo_path);
+	return ret;
+}
+
+#else
+
+enum swupd_code third_party_add_main(int UNUSED_PARAM argc, char UNUSED_PARAM **argv)
+{
+	return SWUPD_OK;
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: No this one doesnt need a TODO as it is function definition when we enable-third-party is provided to auto-configure

File: src/3rd_party_internal.h [hunk 30]
```diff
@@ -0,0 +1,98 @@
...
+
+#include "config.h"
+#include "lib/sys.h"
+#include "swupd.h"
+#include <errno.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_internal.h]: Thanks . Makes sense will remove it in next push

File: src/3rd_party_internal.h [hunk 31]
```diff
@@ -0,0 +1,98 @@
...
+
+	if (fd < 0) {
+		return -1;
+	}
+	close(fd);
+
+	return SWUPD_OK;
+}
+
+static inline int setup_repo_config()
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_internal.h]: Thanks. Makes sense will remove it in next push

File: src/3rd_party_internal.h [hunk 32]
```diff
@@ -0,0 +1,98 @@
...
+{
+	char *repo_config_file;
+	int ret;
+	repo_config_file = sys_path_join(globals.state_dir, "/3rd_party/repo.ini");
+
+	sys_file_exists(repo_config_file);
+
+	if (create_repo_config(repo_config_file)) {
+		debug("Issue creating repo config file: %s in %s\n", repo_config_file, globals.state_dir);
+		goto finish;
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_internal.h]: Makes sense will remove it in next push

File: src/3rd_party_internal.h [hunk 33]
```diff
@@ -0,0 +1,98 @@
...
+	close(fd);
+
+	return SWUPD_OK;
+}
+
+static inline int setup_repo_config()
+{
+	char *repo_config_file;
+	int ret;
+	repo_config_file = sys_path_join(globals.state_dir, "/3rd_party/repo.ini");
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_internal.h]: Okay. Makes sense will remove it in next push

File: src/3rd_party_internal.h [hunk 34]
```diff
@@ -0,0 +1,98 @@
...
+	return;
+}
+
+static inline void remove_repo(char UNUSED_PARAM *repo_name)
+{
+	/* TODO Implement */
+	return;
+}
+
+static inline char *get_repos()
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_internal.h]: Makes sense will remove it in next push

File: src/3rd_party_list.c [hunk 35]
```diff
@@ -0,0 +1,106 @@
...
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#define _GNU_SOURCE
+#include "3rd_party_internal.h"
+#include "config.h"
+#include "lib/log.h"
+#include "swupd.h"
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_list.c]: Makes sense will change it in next push

File: src/3rd_party_remove.c [hunk 36]
```diff
@@ -0,0 +1,110 @@
...
+#include "3rd_party_internal.h"
+#include "config.h"
+#include "lib/log.h"
+#include "swupd.h"
+
+#ifdef THIRDPARTY
+
+int remove_repo_dir(char UNUSED_PARAM *repo_path)
+{
+	//rm_rf(repo_path);
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_remove.c]: No, will remove and add a TODO

File: src/3rd_party_remove.c [hunk 37]
```diff
@@ -0,0 +1,110 @@
...
+	return true;
+}
+
+enum swupd_code third_party_remove_main(int argc, char **argv)
+{
+	enum swupd_code ret = SWUPD_OK;
+	const int step_in_third_party_remove = 1;
+	char *repo_path = NULL;
+
+	/* there is no need to report in progress for init at this time */
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_remove.c]: Makes sense will change it in next push

File: src/swupd_internal.h (hunk 22 above)
Comment (karthikprabhuvinod) [reply src/swupd_internal.h]: Marking this as resolved.

File: src/3rd_party_add.c (hunk 29 above)
[the remark this answers is not in the document]
Comment (otaviobp) [reply src/3rd_party_add.c]: You don't need this function here. REmove it and leave everything related to 3rd party inside the #ifdef

File: src/3rd_party.h [hunk 38]
```diff
@@ -0,0 +1,44 @@
...
+#endif
+
+/**
+ * @brief Creates a new third-party repo under THIRDPARTY_REPO_PREFIX
+ * if it does not exist
+ *
+ * @return 0 on success, SWUPD_THIRD_PARTY_REPO_EXISTS if it already exists
+ * and negative numbers on errors
+ */
+enum swupd_code third_party_add_main(int argc, char **argv);
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party.h]: Thanks for that explanation. Makes sense

**Comment by karthikprabhuvinod** [comment]
> @karthikprabhuvinod Can you remove all functions from this PR that aren't related to the skeleton? And you can add later when you really need them.

Okay done

File: src/3rd_party_add.c (hunk 29 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: Done

File: src/3rd_party_add.c (hunk 27 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: Removed and added a TODO

File: src/3rd_party_list.c [hunk 39]
```diff
@@ -0,0 +1,106 @@
...
+
+#ifdef THIRDPARTY
+
+static void print_repo_list(char UNUSED_PARAM *repos)
+{
+	return;
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_list.c]: Removed  and replaced with TODO

File: src/3rd_party_remove.c [hunk 40]
```diff
@@ -0,0 +1,110 @@
...
+
+	if (ind < 0) {
+		return false;
+	}
+
+	if (argc == 1) {
+		error("The positional args: repo-name is missing\n\n");
+		return false;
+	}
+
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_remove.c]: for above: https://github.com/clearlinux/swupd-client/pull/1161#discussion_r340222876

If think I can do this. I can pre define the no of args expected for each operation (**EXACT_ARG_COUNT**).

Fo eg: for add ...that is 2 (repo-url and repo-name) for remove it is 1(repo-name)

Lets take add as an example:

if (argc - ind) == EXACT_ARG_COUNT

I always check if argc - opind == EXACT_ARG_COUNT..if not fail

File: src/3rd_party_add.c (hunk 26 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: Makes sense.. will change it in next push

File: src/3rd_party_add.c (hunk 10 above)
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: Will address this in a follow-on PR as discussed

File: src/main.c [hunk 41]
```diff
@@ -49,7 +43,7 @@ static enum swupd_code external_search_main(int argc, char **argv)
 
 // TODO(castulo): remove the superseded command from the help menu by end of November 2019,
 // so it is not visible but the command must remain available in the back so we don't break users
-static struct subcmd commands[] = {
+struct subcmd main_commands[] = {
```
Comment (otaviobp) [inline src/main.c]: I realized that now. This should remain static.

**Comment by karthikprabhuvinod** [comment]
> > ignore
> 
> What should I ignore?

I accidently reviewed my own PR..so had to paste some comment

File: src/3rd_party.c [hunk 42]
```diff
@@ -0,0 +1,110 @@
...
+ */
+
+#define _GNU_SOURCE
+#include "config.h"
+#include "swupd.h"
+#include "swupd_internal.h"
+
+#ifdef THIRDPARTY
+
+struct subcmd third_party_commands[] = {
```
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party.c]: same as discussed in
https://github.com/clearlinux/swupd-client/pull/1161#discussion_r340778718

File: src/main.c (hunk 41 above)
Comment (otaviobp) [reply src/main.c]: No. static means that the variable isn't going to be visible outside of this C source file.Using it as a variable is valid and it's ok. Doing that prevents another source file to do this:

```
extern struct subcmd main_commands[];
main_commands = NEW_VALUE;
```

This would change the value of the original main_commands variable

Comment (karthikprabhuvinod) [reply src/main.c]: @otaviobp Thanks for explaining that. I know why it works...subcmd_index is inline static fucntion and it isnt considered another file.

Comment (karthikprabhuvinod) [reply src/main.c]: Fixed this in last push

File: src/3rd_party.c (hunk 42 above)
[the remark this answers is not in the document]
Comment (karthikprabhuvinod) [reply src/3rd_party.c]: Fixed this in last push

**Comment by karthikprabhuvinod** [comment]
@otaviobp @castulo Looks like all comments are addressed. I think it looks ready now

File: src/3rd_party_add.c [hunk 43]
```diff
@@ -0,0 +1,94 @@
...
+		error("The positional args: repo-name and URL are missing\n\n");
+		return false;
+	} else if (argc == 2) {
+		error("The positional args: repo-URL is missing\n\n");
+		return false;
+	}
+
+	/* Ensure that repo add only expects only two args: repo-name, repo-url */
+	if ((argc - ind) != EXACT_ARG_COUNT) {
+		error("Too many Unexpected args\n\n");
```
Comment (castulo) [inline src/3rd_party_add.c]: nitpick, "Too many Unexpected args" sounds redundant, use "Too many arguments" or "Unexpected arguments", also you are using a capital letter in Unexpected in the middle of the sentence

Comment (castulo) [reply src/3rd_party_add.c]: Actually to be consistent with other swupd commands it is better to go with `error("unexpected arguments\n\n");`.

File: src/3rd_party_list.c [hunk 44]
```diff
@@ -0,0 +1,83 @@
...
+static bool parse_options(int argc, char **argv)
+{
+	int ind = global_parse_options(argc, argv, &opts);
+
+	if (ind < 0) {
+		return false;
+	}
+
+	if (argc > ind) {
+		error("Too many Unexpected args\n\n");
```
Comment (castulo) [inline src/3rd_party_list.c]: same here -> `error("unexpected arguments\n\n");`

File: src/swupd_internal.h [hunk 45]
```diff
@@ -22,6 +25,59 @@ enum swupd_code repair_main(int argc, char **argv);
...
+enum swupd_code third_party_main(int argc, char **argv);
+
+/**
+ * @brief Creates a new third-party repo under THIRDPARTY_REPO_PREFIX
+ * if it does not exist
+ *
+ * @return 0 on success, SWUPD_THIRD_PARTY_REPO_EXISTS if it already exists
+ * and negative numbers on errors
+ */
+extern enum swupd_code third_party_add_main(int argc, char **argv);
```
Comment (castulo) [inline src/swupd_internal.h]: I had mentioned this in a previous comment but I could not find it anymore, but you cannot use `enum swupd_code` and return negative numbers on errors. The whole purpose of using `enum soup_code` is to guarantee the function is going to return SWUPD_OK or SWUPD_NO if successful, and any other error within the swupd_exit_codes on errors.

Comment (castulo) [reply src/swupd_internal.h]: found the comment I'd made, here it is:

"All these functions are defined as returning a value of enum swupd_code, which at the moment are positive values from 0 to 36.

When defining functions within the swupd code it is possible to define functions that return any kind of value, boolean, positive numbers, negative numbers, etc. However whenever swupd exits, it must always exit with a positive value contained in swupd_code. In other words all functions that return values different than swupd_code can only be used internally, and whatever function that is going to propagate its return code all the way to the user has to return a value defined within swupd_code.

The whole purpose of using the enum swupd_code as return type is to "certify" in a way that a certain function complies with only returning positive values within the range of swupd_code, and this way we know it is safe to propagate its return code all the way to the user. So by this definition it makes no sense to define a function that supposedly complies with only returning values within swupd_code range, but arbitrarily returning negative numbers on errors. If a function returns a swupd_code then it must return SWUPD_OK or SWUPD_NO in a successful operation, and an appropriate error from swupd_code on failures."

Comment (karthikprabhuvinod) [reply src/swupd_internal.h]: Sorry about missing that. Will correct it in next push

File: src/3rd_party_list.c (hunk 44 above)
Comment (karthikprabhuvinod) [reply src/3rd_party_list.c]: Sure thing Will correct it in next push

File: src/3rd_party_add.c (hunk 43 above)
Comment (karthikprabhuvinod) [reply src/3rd_party_add.c]: Sure thing Will correct it in next push
